arch/rv32i/src/syscall.rs

```diff
@@ -0,0 +1,389 @@
+//! Kernel-userland system call interface for RISC-V architecture.
+
+use core::fmt::Write;
+
+use kernel;
+
+/// This holds all of the state that the kernel must keep for the process when
+/// the process is not executing.
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+pub struct RiscvimacStoredState {
+    /// Store all of the app registers.
+    regs: [usize; 31],
+
+    /// This holds the PC value of the app when the exception/syscall/interrupt
+    /// occurred. We also use this to set the PC that the app should start
+    /// executing at when it is resumed/started.
+    pc: usize,
+
+    /// We need to store the mcause CSR between when the trap occurs and after
+    /// we exit the trap handler and resume the context switching code.
+    mcause: usize,
+}
+
+/// Implementation of the `UserspaceKernelBoundary` for the RISC-V architecture.
+pub struct SysCall();
```

Comment:
I would prefer `pub struct SysCall;` over `empty unit struct` but it is ok.

Reply:
Actually, I think none of the above (including what’s in the code) is what we want. All of these options allow anyone to construct a `SysCall` since all of it’s fields (of which there are none) are public.

`SysCall` needs to be a singleton (I think?), and should only be insatiable through the `unsafe fn new()` function.

The declaration should, therefore, be:

```rust
pub struct SysCall( () ); // Can elide spaces
```

Which would result in an error if someone outside of this module tries to create one. Unlike:

```rust
pub struct SysCall(pub ());
```

Which _can_ be instantiated by anyone.

Reply:
Right, if it needs to be a singleton then it makes sense.

Reply:
I think I updated this correctly, let me know if not.